pytorch_lightning/utilities/auto_restart.py

```diff
@@ -22,12 +22,29 @@
 import numpy as np
 import torch
 from torch.utils.data import Dataset, get_worker_info, Sampler
-from torch.utils.data.dataloader import _MultiProcessingDataLoaderIter, DataLoader, IterableDataset
+from torch.utils.data.dataloader import (
+    _BaseDataLoaderIter,
+    _MultiProcessingDataLoaderIter,
+    _SingleProcessDataLoaderIter,
+    DataLoader,
+    IterableDataset,
+)
 
 import pytorch_lightning as pl
 from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
-from pytorch_lightning.utilities.imports import _fault_tolerant_training
+from pytorch_lightning.utilities.imports import _fault_tolerant_training, _fault_tolerant_training_mode
+
+
+class _FaulTolerantAbstract(abc.ABC):
```

Comment:
Unsure what @awaelchli or @carmocca think but my opinion is that this is confusing. We have an API being enforced internally that the user has to be aware of, but there is no way to know this other than documentation.

I think it's better we take the OO approach by enforcing the user to inherit from a particular class to make the sampler/dataloader stateful (either via inheritance or multiple inheritance)

Reply:
Changed to functional API.